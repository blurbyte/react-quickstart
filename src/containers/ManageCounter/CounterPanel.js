//Presentational (dumb) component

//declared as stateles function
//no need for this keyword, prevents from using react lifecycle methods
//example of ES2015 default parameter and template strings
//doesn't have to alwas receive props: const PresentationalComponent = () => (...);

import React from 'react';
import PropTypes from 'prop-types';

//components
import Wrapper from './Wrapper';
import CounterInfo from './CounterInfo';
import Description from './Description';
import CounterValue from './CounterValue';
import CounterButton from './CounterButton';

const CounterPanel = ({ counter, onIncrease, onDecrease, lowerLimit }) => (
  // disable CounterButton if counter value is lower than limit
  <Wrapper>
    <CounterInfo>
      <Description>Current Counter Value</Description><CounterValue>{counter}</CounterValue>
    </CounterInfo>
    <CounterButton onClick={onIncrease}>Increase Value</CounterButton>
    {
      counter > lowerLimit ?
        <CounterButton onClick={onDecrease}>Decrease Value </CounterButton> :
        <CounterButton disabled>Decrease Value</CounterButton>
    }
  </Wrapper>
);

//list of react PropTypes: https://facebook.github.io/react/docs/typechecking-with-proptypes.html
//you can pass strings, numbers, objects, arrays, etc.
CounterPanel.propTypes = {
  counter: PropTypes.number.isRequired,
  lowerLimit: PropTypes.number.isRequired,
  onIncrease: PropTypes.func,
  onDecrease: PropTypes.func
};

export default CounterPanel;
