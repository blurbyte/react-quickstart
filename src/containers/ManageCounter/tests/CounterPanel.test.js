import React from 'react';
import renderer from 'react-test-renderer';

import CounterPanel from '../CounterPanel';

// mocked components
jest.mock('../Wrapper', () => 'Wrapper');
jest.mock('../CounterInfo', () => 'CounterInfo');
jest.mock('../Description', () => 'Description');
jest.mock('../CounterValue', () => 'CounterValue');
jest.mock('../CounterButton', () => 'CounterButton');

describe('<CounterPanel />', () => {
  it('should render correct button when counter is higher than lowerLimit', () => {
    // props init
    const props = {
      counter: 10,
      onIncrease: jest.fn(),
      onDecrease: jest.fn(),
      lowerLimit: 0
    };
    // component init
    const component = renderer.create(
      <CounterPanel {...props} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
  it('should render disabled button when counter is lower or equal lowerLimit', () => {
    // props init
    const props = {
      counter: 0,
      onIncrease: jest.fn(),
      onDecrease: jest.fn(),
      lowerLimit: 0
    };
    // component init
    const component = renderer.create(
      <CounterPanel {...props} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
