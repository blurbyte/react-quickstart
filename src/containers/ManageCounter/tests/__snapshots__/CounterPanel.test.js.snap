// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CounterPanel /> should render correct button when counter is higher than lowerLimit 1`] = `
<Wrapper>
  <CounterInfo>
    <Description>
      Current Counter Value
    </Description>
    <CounterValue>
      10
    </CounterValue>
  </CounterInfo>
  <CounterButton
    onClick={[Function]}
  >
    Increase Value
  </CounterButton>
  <CounterButton
    onClick={[Function]}
  >
    Decrease Value 
  </CounterButton>
</Wrapper>
`;

exports[`<CounterPanel /> should render disabled button when counter is lower or equal lowerLimit 1`] = `
<Wrapper>
  <CounterInfo>
    <Description>
      Current Counter Value
    </Description>
    <CounterValue>
      0
    </CounterValue>
  </CounterInfo>
  <CounterButton
    onClick={[Function]}
  >
    Increase Value
  </CounterButton>
  <CounterButton
    disabled={true}
  >
    Decrease Value
  </CounterButton>
</Wrapper>
`;
